### PR TITLE
Sampling and histograms

### PR DESCRIPTION
**Pull request by uwieder**
I made some changes to sampling procedures:
1. fixed the bug that 0 was not sampled in IntSet
2. Made all sampling without replacement
3. Support sampling by rate. First compute the sample size and then sample. 

I also added histograms both 1D and 2D. Each with a heavy version - with buckets and a light version. The 2D light version is called 'HeatMap'.

I had (as usual) some issues with git. I hope I haven't messed things up.

### CODE REVIEW

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                                final IStringConverter converterD1, final IStringConverter converterD2,
+                                final IMembershipSet membershipSet) {
+        if (this.initialized) //a histogram had already been created
+            throw new IllegalAccessError("A histogram cannot be created twice");
+        this.initialized = true;
+        final IRowIterator myIter = membershipSet.getIterator();
+        int currRow = myIter.getNextRow();
+        while (currRow >= 0) {
+            if ((columnD1.isMissing(currRow)) || (columnD2.isMissing(currRow))){
+                if (!columnD1.isMissing(currRow))  //only column 2 is missing
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java]: Maybe you can call isMissing only once for each column and save the result in a temporary bool.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            }
+            currRow = myIter.getNextRow();
+        }
+    }
+
+    public Histogram1DLight getMissingHistogramD1() { return this.histogramMissingD1; }
+
+    public long getSize() { return this.totalsize; }
+
+
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java]: One extra empty line here.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    public long getOutOfRange() { return this.outOfRange; }
+
+    /**
+     * @return the index's bucket or null if not been initialized yet
+     */
+    public long getBucket(final int index1, final int index2) {
+        if (!this.initialized)
+            throw new IllegalArgumentException("bucket not initialized yet");
+        if ((index1 < 0) || (index1 >= this.bucketDescDim1.getNumOfBuckets())
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java]: The bounds checks are redundant: the access below to this.buckets will do them anyway.
I am not sure that the "getBucket" name is appropriate. Maybe getCount?

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1D.java [hunk 4]
```diff
@@ -50,17 +60,29 @@ public void createHistogram(final IColumn column, final IMembershipSet membershi
...
+
+    public int getNumOfBuckets() { return this.bucketDescription.getNumOfBuckets(); }
+
     public long getMissingData() { return this.missingData; }
 
     public long getOutOfRange() { return this.outOfRange; }
 
     /**
-     * @return the index's bucket or null if not been initialized yet
+     * @return the index's bucket or exception if doesn't exist.
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1D.java]: You can remove these bound checks as well.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1D.java [hunk 5]
```diff
@@ -50,17 +60,29 @@ public void createHistogram(final IColumn column, final IMembershipSet membershi
...
      */
     public Bucket1D getBucket(final int index) {
         if ((index < 0) || (index >= this.bucketDescription.getNumOfBuckets()))
             throw new IllegalArgumentException("bucket index out of range");
-        return this.buckets[index];
+        Bucket1D res = this.buckets[index];
+        if (res == null)
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1D.java]: Can you add a @NonNull annotation to the return value?
I don't know if we need this null check either.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1D.java [hunk 6]
```diff
@@ -70,7 +92,10 @@ public Bucket1D getBucket(final int index) {
     public Histogram1D union( @NonNull Histogram1D otherHistogram) {
         if (!this.bucketDescription.equals(otherHistogram.bucketDescription))
             throw new IllegalArgumentException("Histogram union without matching buckets");
+        //if ((!this.initialized) || (!otherHistogram.initialized))
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1D.java]: Please delete commented-out code unless you plan to revive it at some point later (e.g., debugging code).

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1DLight.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    public long getMissingData() { return this.missingData; }
+
+    public long getOutOfRange() { return this.outOfRange; }
+
+    /**
+     * @return the index's bucket or exception if doesn't exist.
+     */
+    public long getBucket(final int index) {
+        if (!this.initialized) //a histogram had not been created
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1DLight.java]: Same with these bound checks.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1DLight.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        this.createHistogram(column, membershipSet.sample(sampleRate), converter);
+    }
+
+    public void createSampleHistogram(final IColumn column, final IMembershipSet membershipSet,
+                                      final IStringConverter converter, double sampleRate, long seed) {
+        this.createHistogram(column, membershipSet.sample(sampleRate, seed), converter);
+    }
+
+    /**
+     *
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1DLight.java]: extra empty comment line

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1DLight.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * @param  otherHistogram with the same bucketDescription
+     * @return a new Histogram which is the union of this and otherHistogram
+     */
+    public Histogram1DLight union( @NonNull Histogram1DLight otherHistogram) {
+        if (!this.bucketDescription.equals(otherHistogram.bucketDescription))
+            throw new IllegalArgumentException("Histogram union without matching buckets");
+  //      if ((!this.initialized) || (!otherHistogram.initialized))
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1DLight.java]: Delete commented-out code.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.hiero.sketch.table.api.IColumn;
+import org.hiero.sketch.table.api.IMembershipSet;
+import org.hiero.sketch.table.api.IRowIterator;
+import org.hiero.sketch.table.api.IStringConverter;
+
+/**
+ * A 2 dimension histogram where each bucket is a Bucket2D object. Designed to be used for visualizations where the
+ * number of buckets is small enough so that the semantics of Bucket2D are useful.
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java]: I would rephrase this comment: we want to know more information about each bucket.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                                      final IStringConverter converterD1, final IStringConverter converterD2,
+                                      final IMembershipSet membershipSet, double sampleRate, long seed) {
+        this.createHistogram(columnD1, columnD2, converterD1, converterD2, membershipSet.sample(sampleRate, seed));
+    }
+
+    public Histogram1D getMissingHistogramD1() { return this.histogramMissingD1; }
+
+    public long getSize() { return this.totalsize; }
+
+
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java]: Extra empty line

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        if (this.initialized) //a histogram had already been created
+            throw new IllegalAccessError("A histogram cannot be created twice");
+        this.initialized = true;
+        for (int i = 0; i < this.bucketDescDim1.getNumOfBuckets(); i++)
+            for (int j = 0; j < this.bucketDescDim2.getNumOfBuckets(); j++)
+                this.buckets[i][j] = new Bucket2D();
+        final IRowIterator myIter = membershipSet.getIterator();
+        int currRow = myIter.getNextRow();
+        while (currRow >= 0) {
+            if ((columnD1.isMissing(currRow)) || (columnD2.isMissing(currRow))) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java]: Save the result to isMissing.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+            }
+            currRow = myIter.getNextRow();
+        }
+    }
+
+    public int getNumOfBucketsD1() { return this.bucketDescDim1.getNumOfBuckets(); }
+
+    public int getNumOfBucketsD2() { return this.bucketDescDim2.getNumOfBuckets(); }
+
+
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram2DHeavy.java]: extra empty line?

File: hieroplatform/src/main/java/org/hiero/sketch/table/FullMembership.java [hunk 14]
```diff
@@ -102,9 +106,35 @@ private IMembershipSet sampleUtil(final Random randomGenerator, final int k) {
         final IntSet s = new IntSet(k);
         for (int i=0; i < k; i++)
             s.add(randomGenerator.nextInt(this.rowCount));
+        while (s.size() < k)
+            s.add(randomGenerator.nextInt(this.rowCount));
         return new SparseMembership(s);
     }
 
+    @Override
+    public IMembershipSet sample(double rate){
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/table/FullMembership.java]: Missing space ){

File: hieroplatform/src/main/java/org/hiero/sketch/table/FullMembership.java [hunk 15]
```diff
@@ -102,9 +106,35 @@ private IMembershipSet sampleUtil(final Random randomGenerator, final int k) {
...
+        return this.sample(getSampleSize(rate, 0, false));
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/table/FullMembership.java]: this.getSampleSize

File: hieroplatform/src/main/java/org/hiero/sketch/table/FullMembership.java [hunk 16]
```diff
@@ -102,9 +106,35 @@ private IMembershipSet sampleUtil(final Random randomGenerator, final int k) {
...
+    }
+
+    @Override
+    public IMembershipSet sample(double rate, long seed){
+        return this.sample(getSampleSize(rate, seed, true), seed);
+    }
+
+    private int getSampleSize(double rate, long seed, boolean useSeed) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/table/FullMembership.java]: Could you push some of this code to the interface as a default method?

File: hieroplatform/src/main/java/org/hiero/sketch/table/LazyMembership.java [hunk 17]
```diff
@@ -120,6 +122,31 @@ public IMembershipSet sample(final int k, final long seed) {
...
+    public IMembershipSet sample(double rate){
+        return this.sample(getSampleSize(rate, 0, false));
+    }
+
+    @Override
+    public IMembershipSet sample(double rate, long seed){
+        return this.sample(getSampleSize(rate, seed, true), seed);
+    }
+
+    private int getSampleSize(double rate, long seed, boolean useSeed) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/table/LazyMembership.java]: Duplicated code; should be in the IMembershipSet interface.

File: hieroplatform/src/main/java/org/hiero/sketch/table/SparseMembership.java [hunk 18]
```diff
@@ -110,6 +112,31 @@ public IMembershipSet sample(final int k, final long seed) {
         return new SparseMembership(this.membershipMap.sample(k, seed, true));
     }
 
+
+    @Override
+    public IMembershipSet sample(double rate){
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/table/SparseMembership.java]: These can also be pushed to the interface.

File: hieroplatform/src/main/java/org/hiero/sketch/table/SparseMembership.java [hunk 19]
```diff
@@ -110,6 +112,31 @@ public IMembershipSet sample(final int k, final long seed) {
...
+        return this.sample(getSampleSize(rate, 0, false));
+    }
+
+    @Override
+    public IMembershipSet sample(double rate, long seed){
+        return this.sample(getSampleSize(rate, seed, true), seed);
+    }
+
+    private int getSampleSize(double rate, long seed, boolean useSeed) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/table/SparseMembership.java]: Ditto.

File: hieroplatform/src/main/java/org/hiero/utils/IntSet.java [hunk 20]
```diff
@@ -150,15 +150,23 @@ public IntSet copy() {
...
         final Random psg;
+        int sampleSize;
         if (useSeed)
             psg = new Random(seed);
         else
             psg = new Random();
         int randomKey = psg.nextInt(this.n);
-
-        for (int samples = 0; samples < k; samples++) {
+        if ((this.containsZero) && (randomKey == 0)) {  //sampling zero is done separately
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/utils/IntSet.java]: one space required after //

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+import org.hiero.sketch.table.api.ContentsKind;
+import org.junit.Test;
+import static junit.framework.Assert.assertEquals;
+
+
+public class HistogramTest {
+    @Test
+    public void testHistogram1D() throws Exception {
+        final int bucketnum = 110;
+        BucketsDescriptionEqSize buckDes = new BucketsDescriptionEqSize(0,100,bucketnum);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: spaces after commas

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+        Histogram1D hist = new Histogram1D(buckDes);
+        DoubleArrayColumn col = DoubleArrayTest.generateDoubleArray(10000);
+        FullMembership fmap = new FullMembership(10000);
+        hist.createHistogram(col, fmap, null);
+        int size = 0;
+        for (int i = 0; i < bucketnum; i++)
+            size += hist.getBucket(i).getCount();
+        assertEquals(size + hist.getMissingData() + hist.getOutOfRange(), 10000);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: Nice test.

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 23]
```diff
@@ -0,0 +1,188 @@
+package org.hiero.sketch;
+
+import org.hiero.sketch.spreadsheet.*;
+import org.hiero.sketch.table.ColumnDescription;
+import org.hiero.sketch.table.DoubleArrayColumn;
+import org.hiero.sketch.table.FullMembership;
+import org.hiero.sketch.table.api.ContentsKind;
+import org.junit.Test;
+import static junit.framework.Assert.assertEquals;
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: I think that framework.Assert is deprecated, please use the other assertEquals.

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+        assertEquals(size + hist2.getMissingData() + hist2.getOutOfRange(), 30000);
+        Histogram1D hist3 = new Histogram1D(buckDes);
+        DoubleArrayColumn col2 = DoubleArrayTest.generateDoubleArray(20000);
+        FullMembership fmap2 = new FullMembership(20000);
+        hist3.createSampleHistogram(col2, fmap2, null, 0.1);
+        size = 0;
+        for (int i = 0; i < bucketnum; i++)
+            size += hist3.getBucket(i).getCount();
+        assertEquals(size + hist3.getMissingData() + hist3.getOutOfRange(), 2000);
+        }
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: Too much space.

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+
+    @Test
+    public void testHistogram1DLight() throws Exception {
+        final int bucketnum = 110;
+        BucketsDescriptionEqSize buckDes = new BucketsDescriptionEqSize(0,100,bucketnum);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: commas

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+import static junit.framework.Assert.assertEquals;
+
+
+public class HistogramTest {
+    @Test
+    public void testHistogram1D() throws Exception {
+        final int bucketnum = 110;
+        BucketsDescriptionEqSize buckDes = new BucketsDescriptionEqSize(0,100,bucketnum);
+        Histogram1D hist = new Histogram1D(buckDes);
+        DoubleArrayColumn col = DoubleArrayTest.generateDoubleArray(10000);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: Please save 10000 into a constant and use it everywhere

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+            size += hist3.getBucket(i).getCount();
+        assertEquals(size + hist3.getMissingData() + hist3.getOutOfRange(), 2000);
+        }
+
+    @Test
+    public void testHistogram1DLight() throws Exception {
+        final int bucketnum = 110;
+        BucketsDescriptionEqSize buckDes = new BucketsDescriptionEqSize(0,100,bucketnum);
+        Histogram1DLight hist = new Histogram1DLight(buckDes);
+        DoubleArrayColumn col = DoubleArrayTest.generateDoubleArray(10000);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: Save the constants!

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+            size += hist2.getBucket(i);
+        assertEquals(size + hist2.getMissingData() + hist2.getOutOfRange(), 30000);
+        Histogram1DLight hist3 = new Histogram1DLight(buckDes);
+        DoubleArrayColumn col2 = DoubleArrayTest.generateDoubleArray(20000);
+        FullMembership fmap2 = new FullMembership(20000);
+        hist3.createSampleHistogram(col2, fmap2, null, 0.1);
+        size = 0;
+        for (int i = 0; i < bucketnum; i++)
+            size += hist3.getBucket(i);
+        assertEquals(size + hist3.getMissingData() + hist3.getOutOfRange(), 2000);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: This is surprising, you always get exactly 2000 items?
I hope you are not paying too much for that; we seldom - if ever - will need this degree of precision.

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+        Histogram2DHeavy hist1 = new Histogram2DHeavy(buckDes1, buckDes2);
+        DoubleArrayColumn col3 = DoubleArrayTest.generateDoubleArray(20000);
+        DoubleArrayColumn col4 = DoubleArrayTest.generateDoubleArray(20000);
+        FullMembership fmap1 = new FullMembership(20000);
+        hist1.createHistogram(col3, col4, null, null, fmap1);
+        size = 0;
+        int size1 = 0;
+        int size2 = 0;
+        for (int i = 0; i < bucketnum; i++)
+            for (int j = 0; j< bucketnum; j++)
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: Spacing around <; commas above; constants.

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 30]
```diff
@@ -0,0 +1,188 @@
...
+        for (int i = 0; i < hist2.getNumOfBucketsD2(); i++)
+            size2 += hist2.getMissingHistogramD2().getBucket(i).getCount();
+        size2 += hist2.getMissingHistogramD2().getOutOfRange();
+
+        assertEquals(size + size1 + size2, 30000);
+        Histogram2DHeavy hist3 = new Histogram2DHeavy(buckDes1, buckDes2);
+        DoubleArrayColumn col5 = DoubleArrayTest.generateDoubleArray(20000);
+        DoubleArrayColumn col6 = DoubleArrayTest.generateDoubleArray(20000);
+        FullMembership fmap2 = new FullMembership(20000);
+        hist3.createSampleHistogram(col5, col6, null, null, fmap2, 0.1);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: No testing for the result of this last call?

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 31]
```diff
@@ -0,0 +1,188 @@
...
+        HeatMap hm2 = hm.union(hm1);
+        assertEquals(hm.getSize() + hm1.getSize(), hm2.getSize());
+    }
+
+    private final ColumnDescription desc = new ColumnDescription("SQRT", ContentsKind.Double, true);
+
+    /**
+     * Generates a double array with every fifth entry missing
+     */
+    public  DoubleArrayColumn generateDoubleArray(final int size, final int skip) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: Don't we already have such code somewhere?

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java [hunk 32]
```diff
@@ -0,0 +1,188 @@
...
+        HeatMap hm1 = new HeatMap(buckDes1, buckDes2);
+        DoubleArrayColumn col3 = DoubleArrayTest.generateDoubleArray(20000);
+        DoubleArrayColumn col4 = DoubleArrayTest.generateDoubleArray(20000);
+        FullMembership fmap1 = new FullMembership(20000);
+        hm1.createSampleHistogram(col3, col4, null, null, fmap1, 0.1);
+        HeatMap hm2 = hm.union(hm1);
+        assertEquals(hm.getSize() + hm1.getSize(), hm2.getSize());
+    }
+
+    private final ColumnDescription desc = new ColumnDescription("SQRT", ContentsKind.Double, true);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: Can you also please add a test for a string column with a suitable converter?
The most interesting will be the one with bucket metadata.

File: hieroplatform/src/test/java/org/hiero/sketch/MembershipTest.java [hunk 33]
```diff
@@ -22,12 +22,16 @@ public void TestFullMembership() {
...
         while (tmp >= 0) {
-            //System.out.print(tmp+",");
             tmp = IT.getNextRow();
         }
         assertEquals(this.size, FM.getSize());
         assertTrue(FM.isMember(7));
         assertFalse(FM.isMember(20));
+        final FullMembership FM1 = new FullMembership(1000);
+        final IMembershipSet FM2 = FM1.sample(0.1);
+        assertEquals(100, FM2.getSize());
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/MembershipTest.java]: This is also surprising. Maybe we want two sampling methods: sampleExact and sampleApproximate?
I don't mind just sampleExact if it does not incur an extra significant cost.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java (hunk 1 above)
Comment (uwieder) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java]: I assumed the compiler does that but maybe I'm optimistic, I'll change the code

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java (hunk 28 above)
Comment (uwieder) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: The sampling is exact. It samples rate*size items without replacement. There is no extra cost doing so and it reduces the variance

File: hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java (hunk 31 above)
Comment (uwieder) [inline hieroplatform/src/test/java/org/hiero/sketch/HistogramTest.java]: We do, but there it is hard coded that every fifth item in the column is missing. I wanted for testing that the two columns have missing items in different rows

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java [hunk 34]
```diff
@@ -45,10 +45,12 @@ public void createHistogram(final IColumn columnD1, final IColumn columnD2,
         final IRowIterator myIter = membershipSet.getIterator();
         int currRow = myIter.getNextRow();
         while (currRow >= 0) {
-            if ((columnD1.isMissing(currRow)) || (columnD2.isMissing(currRow))){
-                if (!columnD1.isMissing(currRow))  //only column 2 is missing
+            boolean isMissingD1 = columnD1.isMissing(currRow);
+            boolean isMissingD2 = columnD2.isMissing(currRow);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java]: In general the compiler does not know whether a function has side-effects, so it cannot do common-subexpression elimination for method calls.
But this is not just about efficiency, it is also helping code readability. It's a win-win.
